Added a wrapping function for the  GetGlyphConfidences() function of …

noahmetzger opened the pull request:
…Tesseract

Signed-off-by: Noah Metzger <noah.metzger@bib.uni-mannheim.de>

noahmetzger commented:
The currently used Tesseract version for the tests is bit older and doesn't support glyph confidences.

sirfz commented:
You'll need to add conditions to apply this code only for the supported versions (in order to maintain backward-compatibility). You can leverage the `TESSERACT_VERSION` Cython compile-time variable to achieve this, see the following for an example: https://github.com/sirfz/tesserocr/blob/master/tesserocr.pyx#L2130

noahmetzger commented:
I added the conditions now and reworked the versioning system a bit. Now it is able to differentiate between the different alpha and beta subversions. But it still fails all checks. The error messages doesnt really look to me like they are related to my changes. Maybe you can have a brief look.

sirfz commented:
From the looks of it, there's still a problem with the versioning system as it's applying 4+ code for v3.4 for example.

sirfz commented:
After taking another look at the changes, you forgot to update the version condition on `OcrEngineMode` in tesseract.pxd

bertsky commented:
What is the current status of this PR? 

If I (build and install a recent `libtesseract.so.4` from tesseract/master and) fetch tesserocr/pull/147 and do a `python setup.py build_ext -I/usr/local/include -L/usr/local/lib`, then I will get a `tesserocr.so` that does _not_ include `GetGlyphConfidences`. So how can I (install and) test this?

sirfz commented:
Yep the crash is unrelated to your PR. I'll take a good look whenever I have some time and merge. Thank you

sirfz commented:
Hi @noahmetzger, It seems like `GetGlyphConfidences` has been removed from 4.0.0 as it no longer exists anywhere in the source code.

noahmetzger commented:
Hi @sirfz, yeah there was a discussion about the name of the feature. It is still there but has another name now. I will update this soon.

noahmetzger commented:
I've uploaded a new Version now with updated names

sirfz commented:
Thanks for your work @noahmetzger, 1 last request. Can you add a test case for this new method if possible? Note that you may need to update the version parser in the test file as well.

noahmetzger commented:
@sirfz done

csaftoiu commented:
Can I see working example code with this? I haven't been able to get it to run. It either returns no choices, or it returns a list of some choices for some letters, but not all, and the list changes as I iterate through the result symbol iterator

stweil commented on tests/test_api.py:
```diff
@@ -157,6 +177,27 @@ def test_word_confidences(self):
+                for timestep in choice:
+                    for alternative in timestep:
+                        self.assertGreaterEqual(alternative[1], 0.0)
+                        self.assertLessEqual(alternative[1], 2.0)
```
@noahmetzger, what was the reason for the upper bound 2.0?

The test fails with the current code:

    FAIL: test_LSTM_choices (tests.test_api.TestTessBaseApi)
    Test GetBestLSTMSymbolChoices.
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/media/home/debian/src/github/OCR-D/ocrd_all/tesserocr/tests/test_api.py", line 196, in test_LSTM_choices
        self.assertLessEqual(alternative[1], 2.0)
    AssertionError: 46.24480056762695 not less than or equal to 2.0

noahmetzger replied:
I think this was part of an older version where the probabilities for alternatives have been between 0 and 1.0 according to their percentage. 2.0 was a special marker for the value which was chosen by tesseract for the final result. But this isn't in the code of tesseract anymore.
